File: open_lock.js
/**
 * @param {string[]} deadends
 * @param {string} target
 * @return {number}
 */
 var openLock = function(deadends, target) {
    function* getNeighbor(idx, currString, steps) {
        for (const x of [-1, 1]) {
            const arr = Array.from(currString, (v, i) => {
                const val = parseInt(v);
                if (i === idx) {
                    const newVal = val + x;
                    if (newVal > 9) return 0;
                    else if (newVal < 0) return 9;
                    return newVal;
                }
                return val;
            });
            yield [arr.join(""), steps + 1];
        }
    }
    
    
    function bfs(queue, deadendSet) {
        const visited = new Set();
        visited.add(queue[0]);
        while (queue.length > 0) {
            const [node, steps] = queue.shift();
            if (deadendSet.has(node)) continue;
            if (node === target) return steps;
            for (let i = 0; i < 4; ++i) {
                for (const neighbor of getNeighbor(i, node, steps)) {
                    if (visited.has(neighbor[0])) continue;
                    queue.push(neighbor);
                    visited.add(neighbor[0]);
                }
            }
        }
        return -1;
    }
    
    let queue = [['0000', 0]];
    let deadendSet = new Set(deadends);
    return bfs(queue, deadendSet);
};

// algomonster solution
const nextDigit = new Map();
for (let i = 0; i < 10; i++) {
    let val = (i + 1) % 10;
    nextDigit.set(i.toString(), val.toString());
}
const prevDigit = new Map();
for (let [key, val] of nextDigit.entries()) {
    prevDigit.set(val, key);
}

function numSteps(targetCombo, trappedCombos) {
    if (targetCombo === "0000") return 0;
    const trappedComboSet = new Set([...trappedCombos]);
    const steps = new Map();
    steps.set("0000", 0);
    const bfsQueue = [];
    bfsQueue.push("0000");
    while (bfsQueue.length > 0) {
        let top = bfsQueue.shift();
        for (let i = 0; i < 4; i++) {
            let newCombo = top.slice(0, i) + nextDigit.get(top.charAt(i)) + top.slice(i + 1);
            if (!trappedComboSet.has(newCombo) && !steps.has(newCombo)) {
                bfsQueue.push(newCombo);
                steps.set(newCombo, steps.get(top) + 1);
                if (newCombo === targetCombo) return steps.get(newCombo);
            }
            newCombo = top.slice(0, i) + prevDigit.get(top.charAt(i)) + top.slice(i + 1);
            if (!trappedComboSet.has(newCombo) && !steps.has(newCombo)) {
                bfsQueue.push(newCombo);
                steps.set(newCombo, steps.get(top) + 1);
                if (newCombo === targetCombo) return steps.get(newCombo);
            }
        }
    }
    return -1;
}

File: python/open_lock.py
from pprint import pprint
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        def get_neighbor(string):
            for i, c in enumerate(string):
                curr = list(string)
                for j in [-1, 1]:
                    currDigit = int(c)
                    nextDigit = currDigit + j
                    replaceChar = str(nextDigit % 10)
                    if c != replaceChar:
                        curr[i] = replaceChar
                        yield "".join(curr)
        
        queue = deque([("0000", 0)])
        visited = set()
        while queue:
            node, moves = queue.popleft()
            if node == target:
                return moves
            if node in deadends:
                    continue 
            for neighbor in get_neighbor(node):
                if neighbor in visited:
                    continue
             
                queue.append((neighbor, moves + 1))
                visited.add(neighbor)
        return -1

# leetcode solution
class Solution(object):
    def openLock(self, deadends, target):
        def neighbors(node):
            for i in xrange(4):
                x = int(node[i])
                for d in (-1, 1):
                    y = (x + d) % 10
                    yield node[:i] + str(y) + node[i+1:]

        dead = set(deadends)
        queue = collections.deque([('0000', 0)])
        seen = {'0000'}
        while queue:
            node, depth = queue.popleft()
            if node == target: return depth
            if node in dead: continue
            for nei in neighbors(node):
                if nei not in seen:
                    seen.add(nei)
                    queue.append((nei, depth+1))
        return -1

# other solution
from collections import deque
from typing import List

next_digit = {**{str(i): str(i + 1) for i in range(9)}, "9": "0"}
prev_digit = {e: n for n, e in next_digit.items()}

def num_steps(target_combo: str, trapped_combos: List[str]) -> int:
    if target_combo == "0000":
        return 0
    trapped_combo_set = set(trapped_combos)
    steps = {
        "0000": 0
    }
    bfs_queue = deque({"0000"})
    while bfs_queue:
        top = bfs_queue.popleft()
        for i in range(4):
            new_combo = top[0:i] + next_digit[top[i]] + top[i + 1:]
            if new_combo not in trapped_combo_set and new_combo not in steps:
                bfs_queue.append(new_combo)
                steps[new_combo] = steps[top] + 1
                if new_combo == target_combo:
                    return steps[new_combo]
            new_combo = top[0:i] + prev_digit[top[i]] + top[i + 1:]
            if new_combo not in trapped_combo_set and new_combo not in steps:
                bfs_queue.append(new_combo)
                steps[new_combo] = steps[top] + 1
                if new_combo == target_combo:
                    return steps[new_combo]
    return -1